cmd/zed/zed.d/statechange-blinkled.sh

```diff
@@ -0,0 +1,83 @@
+#!/bin/bash
+#
+# Turn off/on the VDEV's enclosure fault LEDs when the pool's state changes.
+#
+# Turn LED on if the VDEV becomes faulted/degraded, and turn it back off when
+# it's healthy again.
+#
+# Script does nothing if it's unable to find the LED sysfs entries for the VDEV.
+#
+# This script requires ZFS to be built with libdevmapper support.
+#
+[ -f "${ZED_ZEDLET_DIR}/zed.rc" ] && . "${ZED_ZEDLET_DIR}/zed.rc"
+. "${ZED_ZEDLET_DIR}/zed-functions.sh"
+
+# ZEVENT_VDEV_UPATH will not be present if ZFS is not built with libdevmapper
+[ -n "${ZEVENT_VDEV_UPATH}" ] || exit 9
+
+if [ "${ZED_USE_ENCLOSURE_LEDS}" != "1" ] ; then
+	exit 0
+fi
+
+if [ ! -d /sys/class/enclosure ] ; then
+	exit 0
+fi
+
+# Turn on/off enclosure LEDs
+function led
+{
+	name=$1
+	val=$2
+
+	# We want to check the current state first, since writing to the
+	# 'fault' entry always always causes a SES command, even if the
+	# current state is already what you want.
+	if [ -e /sys/block/$name/device/enclosure_device*/fault ] ; then
+
+		# We have to do some monkey business to deal with spaces in
+		# enclosure_device names.  I've seen horrible things like this: 
+		#
+		# '/sys/block/sdfw/device/enclosure_device:SLOT 43 41  /fault'
+		#
+		# ...so escape all spaces.
+		file=`ls /sys/block/$name/device/enclosure_device*/fault | sed 's/\s/\\ /g'`
+
+		current=`cat "$file"`
+
+		# On some enclosures if you write 1 to fault, and read it back,
+		# it will return 2.  Treat all non-zero values as 1 for
+		# simplicity.
+		if [ "$current" != "0" ] ; then
+			current=1
+		fi
+
+		if [ "$current" != "$val" ] ; then
+			# Set the value twice.  I've seen enclosures that were
+			# flakey about setting it the first time.
+			echo $val > "$file"
+			echo $val > "$file"
```

Comment:
Reading the sysfs value is only slightly faster that writing it, at least on our enclosures, so it just makes sense to write it twice. (both operations take in the 100's of milliseconds range).  I'm also not 100% sure that reading the value immediately after writing it would show the correct value, given how flaky they can be.  I haven't done any experimentation on this though, nor have I tested with delays.  I've just seen cases where I've written all the LEDs on all the drives in an enclosure, and one of the LEDs didn't go on.  In those cases, when I wrote the LED again it went on.